lib/lookup/src/lookup_buf/mod.rs

```diff
@@ -0,0 +1,399 @@
+        let mut peeker = self.segments.iter().peekable();
+        let mut next = peeker.next();
+        let mut maybe_next = peeker.peek();
+        while let Some(segment) = next {
+            match segment {
+                SegmentBuf::Field(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"{}."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "{}", segment)?,
+                },
+                SegmentBuf::Index(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"[{}]."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "[{}]", segment)?,
+                },
+                SegmentBuf::Coalesce(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"{}."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "{}", segment)?,
+                },
+            }
+            next = peeker.next();
+            maybe_next = peeker.peek();
+        }
+        Ok(())
+    }
+}
+
+impl LookupBuf {
+    /// Creates an lookup to the root
+    pub fn root() -> Self {
+        Self {
+            segments: VecDeque::new(),
+        }
+    }
+
+    /// Get from the internal list of segments.
+    #[instrument(level = "trace")]
+    pub fn get(&mut self, index: usize) -> Option<&SegmentBuf> {
+        self.segments.get(index)
+    }
+
+    /// Push onto the internal list of segments.
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_back(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_back(segment.into());
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_back(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_back()
+    }
+
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_front(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_front(segment.into())
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_front(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_front()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn iter(&self) -> std::collections::vec_deque::Iter<'_, SegmentBuf> {
+        self.segments.iter()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn len(&self) -> usize {
+        self.segments.len()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    #[instrument(level = "trace")]
+    pub fn is_root(&self) -> bool {
+        self.is_empty()
+    }
+
+    /// Raise any errors that might stem from the lookup being invalid.
+    #[instrument(level = "trace")]
+    pub fn is_valid(&self) -> Result<(), LookupError> {
+        Ok(())
+    }
```

Comment:
Is there a need for this method, if a `LookupBuf` can never be invalid?

Reply:
Nope!